File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <bits/stdc++.h>

using namespace std;

int T, N, M, K;
int a[52][52], visited[52][52];
const int dy[] = { -1, 0, 1, 0 };
const int dx[] = { 0, 1, 0, -1 };

void DFS(int y, int x) {
	visited[y][x] = 1;
	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny < 0 || nx < 0 || ny >= N || nx >= M) continue;
		if(a[ny][nx] == 1 && !visited[ny][nx]) DFS(ny, nx);
	}
	return;
}

int main() {
	cin >> T;
	while (T--) {
		cin >> N >> M >> K;
		memset(a, 0, sizeof(a));
		memset(visited, 0 , sizeof(a));
		int ret = 0;
		for (int i = 0; i < K; i++) {
			int y, x;
			cin >> y >> x;
			a[y][x] = 1;
		}


		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (!visited[i][j] && a[i][j] == 1) {
					ret++;
					DFS(i, j);
				}
			}
		}
		cout << ret << '\n';
	}
	return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int[][] maps;
    static int[][] visited;
    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};

    static void dfs(int y, int x) {
        visited[y][x] = 1;
        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny < 0 || nx < 0 || ny >= maps.length || nx >= maps[0].length || visited[ny][nx] == 1) {
                continue;
            }
            if (maps[ny][nx] == 1) {
                dfs(ny, nx);
            }

        }
    }


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        int testCase = Integer.parseInt(bf.readLine());

        while (testCase-- > 0) {
            StringTokenizer st = new StringTokenizer(bf.readLine());
            int X = Integer.parseInt(st.nextToken());
            int Y = Integer.parseInt(st.nextToken());
            int cabbageCnt = Integer.parseInt(st.nextToken());
            maps = new int[Y][X];
            visited = new int[Y][X];

            for (int i = 0; i < cabbageCnt; i++) {
                StringTokenizer line = new StringTokenizer(bf.readLine());
                int x = Integer.parseInt(line.nextToken());
                int y = Integer.parseInt(line.nextToken());
                maps[y][x] = 1;
            }
            int ret = 0;
            for (int i = 0; i < Y; i++) {
                for (int j = 0; j < X; j++) {
                    if (maps[i][j] == 1 && visited[i][j] == 0) {
                        dfs(i, j);
                        ret++;
                    }
                }
            }

            System.out.println(ret);
        }
    }
}